hopsworks-common/src/main/java/io/hops/hopsworks/common/security/ServiceJWTKeepAlive.java

```diff
@@ -16,77 +16,140 @@
-import org.apache.http.client.methods.HttpGet;
-import org.apache.http.client.protocol.HttpClientContext;
-import org.apache.http.client.utils.URIBuilder;
-import org.apache.http.impl.client.CloseableHttpClient;
-import org.apache.http.impl.client.HttpClients;
-import org.apache.http.protocol.HttpContext;
+import io.hops.hopsworks.jwt.JWTController;
+import io.hops.hopsworks.jwt.exception.JWTException;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.util.BackOff;
+import org.apache.hadoop.util.ExponentialBackOff;
 
 import javax.annotation.PostConstruct;
+import javax.annotation.Resource;
 import javax.ejb.DependsOn;
 import javax.ejb.EJB;
-import javax.ejb.Schedule;
 import javax.ejb.Singleton;
 import javax.ejb.Startup;
-import java.io.IOException;
-import java.net.URI;
-import java.net.URISyntaxException;
+import javax.ejb.Timeout;
+import javax.ejb.TimerConfig;
+import javax.ejb.TimerService;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import java.security.NoSuchAlgorithmException;
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 @Singleton
 @Startup
 @DependsOn("Settings")
+@TransactionAttribute(TransactionAttributeType.NEVER)
 public class ServiceJWTKeepAlive {
-
+  
+  private final static Logger LOGGER = Logger.getLogger(ServiceJWTKeepAlive.class.getName());
+  private final static List<String> SERVICE_RENEW_JWT_AUDIENCE = new ArrayList<>(1);
+  static {
+    SERVICE_RENEW_JWT_AUDIENCE.add("services");
+  }
+  
   @EJB
   private Settings settings;
+  @EJB
+  private JWTController jwtController;
+  @Resource
+  private TimerService timerService;
 
-  private CloseableHttpClient httpClient = HttpClients.createDefault();
-  private URI caURI = null;
-
-  private final static String CA_PATH = "hopsworks-ca/v2/token";
-  private final static Logger LOGGER = Logger.getLogger(ServiceJWTKeepAlive.class.getName());
-
+  private String hostname;
+  private BackOff backOff;
+  
   @PostConstruct
+  @TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
   public void init() {
-    try {
-      caURI = new URIBuilder(settings.getRestEndpoint())
-        .setPath(CA_PATH)
+    hostname = "hopsworks";
+    backOff = new ExponentialBackOff.Builder()
+        .setInitialIntervalMillis(100L)
+        .setMaximumIntervalMillis(2000L)
+        // We issue five renewal tokens so we have four retries
+        .setMaximumRetries(4)
         .build();
-    } catch (URISyntaxException e) {
-      LOGGER.log(Level.SEVERE, "Could not build the URI for the CA", e);
-    }
+    
+    long interval = Math.max(1000L, settings.getServiceJWTLifetimeMS() / 2);
+    timerService.createIntervalTimer(5000L, interval, new TimerConfig("Service JWT renewer", false));
```

Comment:
I think this is optimistic. The token cannot be renewed if it hasn't expired and it lasts 1week by default. If Hopsworks is restarted 3/4 through the token lifetime, we will miss the renewal.

Reply:
You don't account for the leeway period. But in any case I will set it lower

Reply:
I'm not following you. The leeway in the AuthService is set to 0. In any case, even if it was set to 3 days, you would still need to go though this method before any other request using the old token would succeeds.